GpsDataAnalyzer/calculator/alignment_algorithms.py

```diff
@@ -0,0 +1,293 @@
+from datetime import datetime, timedelta
+import numpy as np
+import time
+
+from GpsDataAnalyzer import utils
+
+def find_lineup(set1, set2):
+    """
+    Optimized algorithm to find the approximate offset between two GPS data sets.
+
+    This algorithm first identifies a primary data set and a secondary one based
+    on which starts later. The offset is then applied to the secondary data set.
+    After mapping the timestamps to the corresponding points from each set, it
+    finds the range that has the most overlapping timestamps between the two
+    sets and then finds the optimal offset using the middle of that range 
+    as the starting point.
+
+    Args:
+        set1: GpsDataSet object
+        set2: GpsDataSet object
+
+    Returns:
+        Tuple of two datetimes, the start time for set 1 and for set 2 given the
+        calculated offset. If no lineup is found, it will return None.
+    """
+
+    # find primary data set (the one with the later start time)
+    set1_start_time = set1.gps_data_list[0].time
+    set2_start_time = set2.gps_data_list[0].time
+    set1_end_time = set1.gps_data_list[-1].time
+    set2_end_time = set2.gps_data_list[-1].time 
+    if set1_start_time > set2_start_time:
+        later_start_time = utils.round_time(set1_start_time)
+        primary_set = set1
+        secondary_set = set2
+    else:
+        later_start_time = utils.round_time(set2_start_time)
+        primary_set = set2
+        secondary_set = set1
+
+    # create dict that maps rounded times to points
+    primary_time_points_mapping = create_time_to_points_mapping(primary_set)
+    secondary_time_points_mapping = create_time_to_points_mapping(secondary_set)
+
+    offset_range_length = 50  # how many offsets to check so, for 50, check offsets (-25,25)
+    point_checking_range_length = 100  # points to check around each offset, for 100, check points (-50,50)
+
+    # span length is total length of the span of points to check around the offsets
+    span_length = offset_range_length + point_checking_range_length
+
+    range_start = utils.round_time(max(set1_start_time, set2_start_time)) - timedelta(seconds=span_length)
+    range_end = utils.round_time(min(set1_end_time, set2_end_time)) + timedelta(seconds=span_length)
+    total_seconds = int((range_end-range_start).total_seconds())
+
+    # TODO(ameles): determine if 50 is just a good number or it should be tied to offset size
+    range_optimization_size = offset_range_length
+
+    # count how many are skipped (not overlapping) in first 50  (offset_range_length) points
```

Comment:
L58-88 should go into its own function, something like:
best_range_start, current_range_skip_count = find_best_data_range(range_optimization_size, ...)
then in the docstring of 'find_best_data_range' explain what best means -- which I think is the least number of missing datapoints

Reply:
not needed for this CL, but I'd like to encourage you to think about ways to optimize this function without calling time not in *time_points_mapping 2*n times where n is total_seconds - span_length in the worst case scenario. maybe add it as a todo?